### PR TITLE
LPS-88994, data definition permissions

### PR DESCRIPTION
**Pull request by rbohl**
https://issues.liferay.com/browse/LPS-88994, ready for review.

### CODE REVIEW

**Comment by rbohl** [comment]
Gabriel, see the last commit for just the permissions draft.

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown [hunk 1]
```diff
@@ -94,10 +105,85 @@ To delete a `DEDataDefinition`, pass a `DEDataDefinitionDeleteRequest` to
...
+- `long scopeGroupId`
+- `String[] roleNames`
+
+Optional parameters: none
+
+Optional builder methods:
+- `allowAddDefinition()` to grant Add permissions
+- `allowDefinePermissions()` to grant permissions to grant permissions
+<!-- Find a better way to say the above -->
+
```
Comment (marcelabc) [inline developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown:158]: @rbohl can you add this phrase right after you specified the optional methods?
"Specify more than one builder method to grant multiple permissions
simultaneously."

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown [hunk 2]
```diff
@@ -94,10 +105,85 @@ To delete a `DEDataDefinition`, pass a `DEDataDefinitionDeleteRequest` to
...
+Required parameters: 
+- `long companyId`
+- `long scopeGroupId`
+- `String[] roleNames`
+
+Optional parameters: none
+
+Optional builder methods:
+- `allowAddDefinition()` to grant Add permissions
+- `allowDefinePermissions()` to grant permissions to grant permissions
```
Comment (marcelabc) [inline developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: this specification could be: 
"allowDefinePermissions(): the user will be allowed to grant permission to other users."
It's complicated to explain that but if you choose not to edit this part, it's okay to me 👍

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown [hunk 3]
```diff
@@ -94,10 +105,85 @@ To delete a `DEDataDefinition`, pass a `DEDataDefinitionDeleteRequest` to
...
+
+Optional builder methods:
+- `allowDelete()` to grant deletion permission
+- `allowUpdate()` to grant update permission
+- `allowView()` to grant view permission
+
+Specify more than one builder method to grant mutliple permissions
+simultaneously.
+
+Here's an example request builder that grants Site Members all three permissions:
```
Comment (marcelabc) [inline developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: where is the example?

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown [hunk 4]
```diff
@@ -94,10 +105,85 @@ To delete a `DEDataDefinition`, pass a `DEDataDefinitionDeleteRequest` to
...
+Optional parameters: none
+
+Once the `DEDataDefinitionDeletePermissionsRequest` is executed, all specified
+Roles lose both permissions simultaneously: Role Users can no long add Data
+Definitions or manage permissions for Data Definitions. 
+
+### Granting and Revoking Permission to View, Update, or Delete a specific Data Definition
+
+`DEDataDefinitionSaveModelPermissionsRequest` is for granting permission to
+update, view, and delete a specific Data Definition already in the system.
```
Comment (marcelabc) [inline developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: update, view, **and/or** delete

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown [hunk 5]
```diff
@@ -94,10 +105,85 @@ To delete a `DEDataDefinition`, pass a `DEDataDefinitionDeleteRequest` to
...
+simultaneously.
+
+Here's an example request builder that grants Site Members all three permissions:
+
+`DEDataDefinitionDeleteModelPermissionsRequest` is for revoking permission to
+view, update, or delete two permissions: the permission to define new Data
+Definition permissions for a User, and the permission to add a new entity
+registry in the system. If a user choose to make this type of delete request,
+all the roles listed in the request will lose both permissions at the same
+time.
```
Comment (marcelabc) [inline developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: This paragraph is not right. The difference of the DeleteModelPermission to the DeletePermission is that the user can specify in the "actionIDs" parameter which permission will revoke, while in the other one the user loses all permissions at the same time without specifying anything.

I wrote in this way, but you can change if you want to:
DEDataDefinitionDeleteModelPermissionsRequest is for revoking permission to view, update, and/or delete a specific data definition already in the system. The user can insert one or more permissions that wants to revoke in the “actionIds” parameter in the request.

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown (hunk 2 above)
Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: @marcelabc can the Role User with this permission grant permission to manage Model Permissions or to manage Add and DefinePermission? Also, do they grant this to a Role?

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown (hunk 3 above)
Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: Oops, I moved it to the top of the article :)

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown (hunk 5 above)
Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: Ahh, my apologies. I copied this into the wrong place and didn't finish editing it. I'll fix it.

File: developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown (hunk 2 above)
Comment (marcelabc) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: The user that has been granted with the permission "DefinePermission" can grant another user with Model Permissions and/or Add/DefinePermission. Besides that, a user can only use the Delete Permission/Model Permission if has the "Define Permission". 

It`s a lot of detail, right?  :)

Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: It is a lot of detail! Let's make sure I have it right:
Let's say we give RoleA the allowDefinePermissions permission but not the allowAddDefinitions permission.

That means a User in RoleA can now
1. Give another Role permission to view, update, or delete a specific data definition
2. Give another Role permission to grant other Roles permissions.
3. Delete a Data Definition IF they also have the delete Model Permission for that defintiionId.

But that User cannot
1. Add Data Definitions.
2. Grant other Roles Add Definitions permission.

Is that right?

Comment (marcelabc) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: You are almost there!
 
User in RoleA can now:


1. Give another Role permission to view, update, or delete a specific data definition 
2. Give another Role permission to DefinePermission
3. Give another Role permission to AddDataDefinition
4. Revoke permission (definePermission/AddDataDefinition)
5. Revoke model permission (delete/update/view)

**another role can be his own role or another one.**

The user in RoleA can only perform action as add/delete/update/view if his role receives the permission.

Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: Okay, I think I have it. I just find it interesting and a little confusing that the User in RoleA cannot Add Data Definitions but can give another Role permission to Add Data Definitions.

Comment (rbohl) [reply developer/frameworks/articles/data-engine/01-data-definition-api/01-data-definition-api-operations.markdown]: Oh I see, so he could grant himself the Add permission too. That sounds fair :)

**Comment by rbohl** [comment]
@marcelabc @gabrielwas see my latest commit which tries to address Marcela's comments. Thanks!

**Comment by marcelabc** [comment]
I reviewed it now, it looks good to me 👍

**Pull request closed without merging**